From 6de3bd1c95180e124086dc3577cfdab527f958b3 Mon Sep 17 00:00:00 2001
From: Jason Wang <jasowang@redhat.com>
Date: Mon, 29 Sep 2014 00:51:32 -0400
Subject: [x86] hyperv: Bypass the timer_irq_works() check

Message-id: <1411951892-53327-1-git-send-email-jasowang@redhat.com>
Patchwork-id: 96743
O-Subject: [RHEL7.1 PATCH] x86, hyperv: Bypass the timer_irq_works() check
Bugzilla: 1058105
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>

Bugzilla: 1058105
Brew Build: https://brewweb.devel.redhat.com/taskinfo?taskID=8036291
Test status: Boot test by myself
Notes: Conflicts because of header inclusion

This patch bypass the timer_irq_works() check for hyperv guest since:

- It was guaranteed to work.
- timer_irq_works() may fail sometime due to the lpj calibration were inaccurate
  in a hyperv guest or a buggy host.

In the future, we should get the tsc frequency from hypervisor and use preset
lpj instead.

[ hpa: I would prefer to not defer things to "the future" in the future... ]

Cc: K. Y. Srinivasan <kys@microsoft.com>
Cc: Haiyang Zhang <haiyangz@microsoft.com>
Cc: <stable@vger.kernel.org>
Acked-by: K. Y. Srinivasan <kys@microsoft.com>
Signed-off-by: Jason Wang <jasowang@redhat.com>
Link: http://lkml.kernel.org/r/1393558229-14755-1-git-send-email-jasowang@redhat.com
Signed-off-by: H. Peter Anvin <hpa@linux.intel.com>
(cherry picked from commit ca3ba2a2f4a49a308e7d78c784d51b2332064f15)
Signed-off-by: Jason Wang <jasowang@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

Conflicts:
 arch/x86/kernel/cpu/mshyperv.c

Signed-off-by: Jason Wang <jasowang@redhat.com>

diff --git a/arch/x86/kernel/cpu/mshyperv.c b/arch/x86/kernel/cpu/mshyperv.c
index d15a4c4..98e770c 100644
--- a/arch/x86/kernel/cpu/mshyperv.c
+++ b/arch/x86/kernel/cpu/mshyperv.c
@@ -25,6 +25,7 @@
 #include <asm/idle.h>
 #include <asm/irq_regs.h>
 #include <asm/i8259.h>
+#include <asm/timer.h>
 
 struct ms_hyperv_info ms_hyperv;
 EXPORT_SYMBOL_GPL(ms_hyperv);
@@ -95,6 +96,11 @@ static void __init ms_hyperv_init_platform(void)
 
  if (ms_hyperv.features & HV_X64_MSR_TIME_REF_COUNT_AVAILABLE)
   clocksource_register_hz(&hyperv_cs, NSEC_PER_SEC/100);
+
+#ifdef CONFIG_X86_IO_APIC
+ no_timer_check = 1;
+#endif
+
 }
 
 const __refconst struct hypervisor_x86 x86_hyper_ms_hyperv = {
-- 
1.7.1